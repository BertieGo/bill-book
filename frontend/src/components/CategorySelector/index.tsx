import React from 'react';
import {Select} from 'antd';
import isFunction from 'lodash/isFunction';
import {ICategorySelectorProps} from '../../declare';
import {MSG_FORM} from '../../constants';
import isEqual from "lodash/isEqual";

const {Option} = Select;

export default class CategorySelector extends React.Component<ICategorySelectorProps, {}> {

    shouldComponentUpdate(nextProps: Readonly<ICategorySelectorProps>, nextState: Readonly<{}>, nextContext: any): boolean {
        const {categories, isFisFetchingBillCategories} = this.props;
        return !isEqual(nextProps.categories, categories) || !isEqual(nextProps.isFisFetchingBillCategories, isFisFetchingBillCategories);
    }

    onChange = (value: string) => {
        const {onChange} = this.props;
        if (isFunction(onChange)) {
            onChange(value);
        }
    }

    render() {
        const {categories, isFisFetchingBillCategories} = this.props;

        return (
            <Select
                style={{width: 120}}
                onChange={this.onChange}
                loading={isFisFetchingBillCategories}
                allowClear
                placeholder={MSG_FORM.CATEGORY_REQUIRED}
            >
                {
                    categories.map((category) => (
                        <Option key={category.id} value={category.id}>{category.name}</Option>
                    ))
                }
            </Select>
        )
    }
}
