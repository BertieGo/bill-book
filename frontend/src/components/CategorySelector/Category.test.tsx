import React from 'react';
import CategorySelector from './index';
import renderer from 'react-test-renderer';
import createReactAppJestPolyfill from '../../utils/test/polyfill';

createReactAppJestPolyfill();

test('COMPONENT-TEST: CategorySelector', () => {
    const data = [{
        id: '8s0p77c323',
        time: "2019-06-30T16:00:00.000Z",
        name: '支出',
        type: 0,
        _id: '5efdc622c42bdc4067a47c5c'
    }];

    const tree = renderer
        .create(<CategorySelector categories={data as any[]} isFisFetchingBillCategories={false} onChange={() => {}}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});


