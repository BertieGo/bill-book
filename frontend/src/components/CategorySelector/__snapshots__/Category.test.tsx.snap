// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT-TEST: CategorySelector 1`] = `
<div
  className="ant-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "width": 120,
    }
  }
>
  <div
    className="ant-select-selector"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <span
      className="ant-select-selection-search"
    >
      <input
        aria-activedescendant="undefined_list_0"
        aria-autocomplete="list"
        aria-controls="undefined_list"
        aria-haspopup="listbox"
        aria-owns="undefined_list"
        autoComplete="off"
        className="ant-select-selection-search-input"
        onChange={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onPaste={[Function]}
        readOnly={true}
        role="combobox"
        style={
          Object {
            "opacity": 0,
          }
        }
        unselectable="on"
        value=""
      />
    </span>
    <span
      className="ant-select-selection-placeholder"
    >
      请选择类型
    </span>
  </div>
  <span
    aria-hidden={true}
    className="ant-select-arrow"
    onMouseDown={[Function]}
    style={
      Object {
        "WebkitUserSelect": "none",
        "userSelect": "none",
      }
    }
    unselectable="on"
  >
    <span
      aria-label="down"
      className="anticon anticon-down"
      role="img"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </span>
  </span>
</div>
`;

exports[`Link renders correctly 1`] = `
<div>
  <div
    className="ant-row mb10"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-4"
      style={Object {}}
    >
      支出
      ：
    </div>
    <div
      className="ant-col ant-col-8"
      style={Object {}}
    >
      <div
        className="ant-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-col-20"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <div
            className="row-bar"
            style={
              Object {
                "backgroundColor": "#72a935",
                "width": "100%",
              }
            }
          />
        </div>
        <div
          className="ant-col ant-col-4"
          style={Object {}}
        >
          <span
            className="ml20"
          >
            2672740
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
